corehq/apps/events/models.py

```diff
@@ -248,6 +277,26 @@ def delete(self, using=None, keep_parents=False):
     def status(self):
         return self.attendee_list_status
 
+    def get_total_attendance_takers(self):
+        return len(self.attendance_taker_ids)
+
+    @property
+    def fake_case_id(self):
+        """
+        A fake case ID, to prevent the Event's case ID clashing with the
+        Event's case sharing group's ID
+        """
+        return self.event_id + '-0'
+
+
+def get_user_case_sharing_groups_for_events(commcare_user):
+    """
+    Creates a case sharing group for every `Event` that the `commcare_user`
+    is an attendance taker for in their domain.
+    """
+    for event in Event.objects.by_domain(commcare_user.domain):
+        if commcare_user.user_id in event.attendance_taker_ids:
+            yield event.get_case_sharing_group(commcare_user.user_id)
```

Comment:
I think the docstring needs an update :thinking:

Reply:
Just noting this here for others that might be as confused as I was when I saw this the usage of `yield` here. Since we're calling this function like 
```python
groups += get_user_case_sharing_groups_for_events(self)
```
I expected it to return only one value, since it looks like we're calling the function one, but `+=` iterates over the given list or generator. I did not know that. I am curious as to why you did it this way, instead of
```python
def get_user_case_sharing_groups_for_events(commcare_user):
    """
    Creates a case sharing group for every `Event` that the `commcare_user`
    is an attendance taker for in their domain.
    """
    groups = []
    for event in Event.objects.by_domain(commcare_user.domain):
        if commcare_user.user_id in event.attendance_taker_ids:
            groups.append(event.get_case_sharing_group(commcare_user.user_id))
    return groups
```

Reply:
Docstring has been updated. :+1:

With regards to your question, I think @kaapstorm would be the best person to answer that as he implemented this during our pairing session.

Reply:
We used `yield` because we were referring to how case sharing groups are implemented for Locations, and [it uses `yield`](https://github.com/dimagi/commcare-hq/blob/db36c7d617756ea5b27e1c360769ead347bd5e4b/corehq/apps/locations/models.py#L789).

> Is there some benefit of using yield that I'm missing?

`yield` is great for anything that you only need to iterate once.

(If you haven't seen them before, you might enjoy David Beazley's series of talks around generators:
1. [Generator Tricks for Systems Programmers](http://www.dabeaz.com/generators/) (PyCon 2008)
2. [A Curious Course on Coroutines and Concurrency](https://www.youtube.com/watch?v=Z_OAlIhXziw) (PyCon 2009)
3. [Generators: The Final Frontier](https://www.youtube.com/watch?v=D1twn9kLmYg) (PyCon 2014)

(I couldn't find a video of the first one, but the slides are there. The second two links go to YouTube.))

Your surprise is well noted though, @SmittieC . Maybe generators should be named `iter_...` instead of `get_...`, as in `def iter_user_case_sharing_groups_for_events(commcare_user)`.

Reply:
Excellent, thanks!